### PR TITLE
test: meta transaction FT transfer

### PR DESCRIPTION
**Pull request by jakmeier**
This is the example used for NEP-366, an FT transfer paid by the
relayer but the relayer gets a reward in FT.

Adds `fungible_token.wasm` as a test contract, which is the
example implementation of NEP-141 in the near-sdk-rs repository.

We deploy this WASM, initialize the state to make an FT transfer
possible and then execute a meta transaction as intended.

This makes gas checks in meta transaction tests even more annoying
but otherwise the existing test framework mostly just worked.

### CODE REVIEW

File: test-utils/testlib/src/runtime_utils.rs [hunk 1]
```diff
@@ -23,28 +21,27 @@ pub fn x_dot_y_dot_alice_account() -> AccountId {
...
 }
 
-static DEFAULT_TEST_CONTRACT_HASH: Lazy<CryptoHash> =
-    Lazy::new(|| hash(near_test_contracts::rs_contract()));
-
 pub fn add_test_contract(genesis: &mut Genesis, account_id: &AccountId) {
+    add_contract(genesis, account_id, near_test_contracts::rs_contract().to_vec())
+}
+
+pub fn add_contract(genesis: &mut Genesis, account_id: &AccountId, code: Vec<u8>) {
```
Comment (mm-near) [inline test-utils/testlib/src/runtime_utils.rs:28]: comment.

(I guess that it adds the contract for this account in the genesis - but please mention explicitly)

File: test-utils/testlib/src/runtime_utils.rs [hunk 2]
```diff
@@ -23,28 +21,27 @@ pub fn x_dot_y_dot_alice_account() -> AccountId {
     "x.y.alice.near".parse().unwrap()
 }
 
-static DEFAULT_TEST_CONTRACT_HASH: Lazy<CryptoHash> =
-    Lazy::new(|| hash(near_test_contracts::rs_contract()));
-
 pub fn add_test_contract(genesis: &mut Genesis, account_id: &AccountId) {
+    add_contract(genesis, account_id, near_test_contracts::rs_contract().to_vec())
```
Comment (mm-near) [inline test-utils/testlib/src/runtime_utils.rs:25]: also comment this function (add_test_contract) -- which contract does it add ?

File: integration-tests/src/tests/client/features/delegate_action.rs [hunk 3]
```diff
@@ -395,3 +416,148 @@ fn meta_tx_delete_account() {
...
+        ft_transfer_event(&sender, &relayer, 10),
+        "relayer event looks wrong"
+    );
+    assert_eq!(
+        fn_call_logs[1],
+        ft_transfer_event(&sender, &receiver, 1000),
+        "receiver event looks wrong"
+    );
+
+    // let relayer_payment_event: serde_json::Value =
```
Comment (mm-near) [inline integration-tests/src/tests/client/features/delegate_action.rs:495]: ? what about this code ?

File: integration-tests/src/tests/client/features/delegate_action.rs [hunk 4]
```diff
@@ -395,3 +416,148 @@ fn meta_tx_delete_account() {
...
+    //     serde_json::from_str(&format!("{{{}}}", fn_call_logs[0]))
+    //         .expect("log events should be JSON");
+    // assert_eq!(relayer_payment_event, ft_transfer_event(&sender, &relayer, 10));
+
+    // let transfer_event: serde_json::Value =
+    //     serde_json::from_str(&fn_call_logs[1]).expect("log events should be JSON");
+    // assert_eq!(transfer_event, ft_transfer_event(&sender, &receiver, 1000));
```
Comment (mm-near) [inline integration-tests/src/tests/client/features/delegate_action.rs:502]: can you also check that david got the tokens ? (and that the token balance of alice & bob are correct)

**Review by mm-near — APPROVED** [review]

File: integration-tests/src/tests/client/features/delegate_action.rs (hunk 3 above)
Comment (jakmeier) [reply integration-tests/src/tests/client/features/delegate_action.rs]: should have been deleted, thanks

File: test-utils/testlib/src/runtime_utils.rs (hunk 2 above)
Comment (jakmeier) [reply test-utils/testlib/src/runtime_utils.rs:29]: done

File: test-utils/testlib/src/runtime_utils.rs (hunk 1 above)
Comment (jakmeier) [reply test-utils/testlib/src/runtime_utils.rs:33]: done

File: integration-tests/src/tests/client/features/delegate_action.rs (hunk 4 above)
Comment (jakmeier) [reply integration-tests/src/tests/client/features/delegate_action.rs]: yes, I added a small helper `assert_ft_balance` and then 

```rust
assert_ft_balance(&node, &ft_contract, &receiver, 1000);
assert_ft_balance(&node, &ft_contract, &sender, 10_000 - 1000 - 10);
assert_ft_balance(&node, &ft_contract, &relayer, 1_000_000 - 10_000 + 10);
```

**Pull request merged**